### ISSUE TITLE
`Aggregate` function not working in combination with QFieldSync and offlineediting

### ISSUE THREAD

**Issue by ghtmtt** [opengisch/QField#734]
If an aggregated function is set in QGIS (like suggested in the doc https://qfield.org/docs/project-management/vector-layers.html?highlight=variables#additional-variables), QField throws an error "layer not found".

### Reproduction Steps:

In attach the test project [pontedera-offline.zip](https://github.com/opengisch/QField/files/3912581/pontedera-offline.zip)

* `base` is the parent layer
* `survey` is the child layer
* `prova` field of `survey` is set with the following expression in QGIS:
```
aggregate(
 layer:= 'base',
 aggregate:='max',
 expression:=id_fermata,
 filter:=intersects(buffer($geometry, 0.0005), geometry(@parent)))
 ```
* when adding a new point in QGIS the `prova` field is updated with the `max` value of the `id_fermata` of the `base` layer
* when the project is packaged with QFieldSync and then loaded in QField, when adding a new point to `survey` the field `prova` is not updated and the error of `not finding base layer` is shown.

### Expected behavior:

Value added in the field

### Observed behavior:

QField error `layer base not found`

**QField version:** 1.2.0 (10200993)

## Additional information:

* Problem can be reliably reproduced, doesn't happen randomly: Yes
* Problem happens with all files and projects, not only some files or projects: Yes

**Comment by m-kuhn** [comment]
Maybe the layer is renamed to `base (offline)` by offline editing?

**Comment by ghtmtt** [comment]
Not in the legend. Even if renamed, wouldn't it cause some trouble when syncing back the project?

**Comment by m-kuhn** [comment]
Hmm... when I open the demo project you uploaded it loads like this:

![image](https://user-images.githubusercontent.com/588407/69983567-531cc380-1537-11ea-9097-01d76affd4ae.png)

**Comment by m-kuhn** [comment]
We should probably add a flag to offline editing to not rename any layers.

**Comment by ghtmtt** [comment]
> Hmm... when I open the demo project you uploaded it loads like this

that's weird. I zipped the project exported from QFieldSync. Let's try with the same project exported by Package Layer. [test.zip](https://github.com/opengisch/QField/files/3912696/test.zip)

**Comment by soester** [comment]
@ghtmtt  if you add ' (offline)' to the layername in your expression as suggested by @m-kuhn qgis will tell you that the experssion is not valid. 
But the aggregate function should work in qfield. We use it a lot this way.

Project configuration would be easier if layernames would not be renamed by qfieldsync.

**Comment by m-kuhn** [comment]
Should we migrate this to a QGIS offline editing issue to make the suffix ` (offline)` optional?

**Comment by 3nids** [comment]
A nice hack for this would be to create a custom expression function called offline_layer_id, taking a layer id as a parameter. In normal project it would return it untouched, and append _offline in the offline project.

**Comment by m-kuhn** [comment]
Yes, that should work.
Also, hacking qfieldsync to strip away the ` (offline)` after offlining would be about the same complexity and will not require in-depth knowledge of the user.

**Comment by ghtmtt** [comment]
> A nice hack for this would be to create a custom expression function called offline_layer_id, taking a layer id as a parameter. In normal project it would return it untouched, and append _offline in the offline project

mmm yeah. But how to distinguish if a layer comes from an offline project?

**Comment by she-weeds** [comment]
I've come up with an interim workaround using `coalesce()` and `layer_property()` to pick the first valid layer name reference out of a potential list in [this GIS.SE answer](https://gis.stackexchange.com/a/360296/98784). You can therefore use `aggregate()` to refer to `layer name` or the `(offline)` variant.

You can use `with_variable()` to avoid having to update any layer name references twice.

If the list gets long and you don't want a whole heap of `layer_property()` functions you could use array functions (probably worth it only for much longer lists as it looks quite ugly too).

For example, a list of potential layer names and `(offline)` variants

```
with_variable('my_layer',
              with_variable('my_array', 
                            array('layer_name',
                                  'another_name', 
                                  'third_name'),
                            array_first(
                                array_filter(
                                    array_foreach(@my_array,
                                                  coalesce(layer_property(@element,'name'),
                                                           layer_property(@element||' (offline)','name))),
                                    @element is not null))),
              case when @my_layer IS NULL 
                   then 'Layer not found' 
                   else 
                   aggregate(layer:= @my_layer,
                             aggregate:='concatenate',
                             expression:=to_string("LAYER_ATTRIBUTE"),
                             concatenator:=', ',
                             filter:=intersects($geometry, geometry(@parent))) 
              end)
```

Example (layer names in array - `road`, `roads`, `tr_road`)

![layer_id3](https://user-images.githubusercontent.com/41975056/80897866-aa220100-8d40-11ea-8d6f-e3120525560c.gif)

Can someone test this using QFieldSync? (I don't currently have access to an android device to test)

**Comment by m-kuhn** [comment]
This should be fixed with qgis 3.16.2 no longer appending the offline suffix?

**Comment by ghtmtt** [comment]
yep, confirmed. Closing the issue. Thanks!
